src/agent/onefuzz-agent/src/tasks/coverage/generic.rs

```diff
@@ -0,0 +1,344 @@
+            self.poller.run(callback).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn load_filter(&self) -> Result<CmdFilter> {
+        let raw_filter_path = if let Some(raw_path) = &self.config.coverage_filter {
+            raw_path
+        } else {
+            return Ok(CmdFilter::default());
+        };
+
+        // Ensure users can locate the filter relative to the setup container.
+        let expand = Expand::new().setup_dir(&self.config.common.setup_dir);
+        let filter_path = expand.evaluate_value(raw_filter_path)?;
+
+        let data = fs::read(&filter_path).await?;
+        let def: CmdFilterDef = serde_json::from_slice(&data)?;
+        let filter = CmdFilter::new(def)?;
+
+        Ok(filter)
+    }
+}
+
+async fn deserialize_or_default<T>(path: impl AsRef<Path>) -> Result<T>
+where
+    T: Default + DeserializeOwned,
+{
+    use tokio::io::ErrorKind::NotFound;
+
+    let data = fs::read(path).await;
+
+    if let Err(err) = &data {
+        if err.kind() == NotFound {
+            return Ok(T::default());
+        }
+    }
+
+    let data = data?;
+
+    Ok(serde_json::from_slice(&data)?)
+}
+
+struct TaskContext<'a> {
+    // Optional only to enable temporary move into blocking thread.
+    cache: Option<ModuleCache>,
+
+    config: &'a Config,
+    coverage: CommandBlockCov,
+    filter: CmdFilter,
+    heartbeat: Option<TaskHeartbeatClient>,
+}
+
+impl<'a> TaskContext<'a> {
+    pub fn new(
+        cache: ModuleCache,
+        config: &'a Config,
+        coverage: CommandBlockCov,
+        filter: CmdFilter,
+        heartbeat: Option<TaskHeartbeatClient>,
+    ) -> Self {
+        let cache = Some(cache);
+
+        Self {
+            cache,
+            config,
+            coverage,
+            filter,
+            heartbeat,
+        }
+    }
+
+    pub async fn record_input(&mut self, input: &Path) -> Result<()> {
+        let coverage = self.record_impl(input).await?;
+        self.coverage.merge_max(&coverage);
+
+        Ok(())
+    }
+
+    async fn record_impl(&mut self, input: &Path) -> Result<CommandBlockCov> {
+        // Invariant: `self.cache` must be present on method enter and exit.
+        let cache = self.cache.take().expect("module cache not present");
+
+        let filter = self.filter.clone();
+        let cmd = self.command_for_input(input)?;
+        let recorded = spawn_blocking(move || record_os_impl(cache, cmd, filter)).await??;
+
+        // Maintain invariant.
+        self.cache = Some(recorded.cache);
+
+        Ok(recorded.coverage)
+    }
+
+    fn command_for_input(&self, input: &Path) -> Result<Command> {
+        let expand = Expand::new()
+            .input_path(input)
+            .job_id(&self.config.common.job_id)
+            .setup_dir(&self.config.common.setup_dir)
+            .target_options(&self.config.target_options)
+            .task_id(&self.config.common.task_id);
+
+        let mut cmd = Command::new(&self.config.target_exe);
```

Comment:
this should use the expansion, such that we can use `{setup_dir}` rather than hard-coding `setup/` at the service level.